root/checkopslevel.go

```diff
@@ -0,0 +1,161 @@
+// LifecycleToLevel maps lifecycle index to level index.
+// We want to keep this at the index level in case names or other attributes change.
+var DefaultLifecycleToLevelMap = LifecycleToLevelMap{
+	DevelopmentLifecycle: BronzeLevel,
+	PrivateBetaLifecycle: SilverLevel,
+	PublicBetaLifecycle:  SilverLevel,
+	PublicLifecycle:      SilverLevel,
+	OpsLifecycle:         SilverLevel,
+	EndOfLifeLifecycle:   BeginnerLevel,
+}
+
+// CheckOpsLevel checks if the app passes all the required OpsLevel checks.
+func CheckOpsLevel(ctx context.Context, appName string) error {
+	client, err := newClient(DefaultLifecycleToLevelMap)
+	if err != nil {
+		return errors.Wrap(err, "creating opslevel client")
+	}
+
+	serviceId, err := client.GetServiceIdWithAlias(appName)
+	if err != nil {
+		return errors.Wrap(err, "getting service id")
+	}
+
+	service, err := client.GetService(serviceId.Id)
+	if err != nil {
+		return errors.Wrap(err, "getting service")
+	}
+
+	sm, err := client.GetServiceMaturityWithAlias(appName)
+	if err != nil {
+		return errors.Wrap(err, "getting maturity reports")
+	}
+
+	isCompliant, err := client.isCompliant(service, sm)
+	if err != nil {
+		return errors.Wrap(err, "checking compliance")
+	}
+
+	if isCompliant {
+		return nil
+	}
+
+	return fmt.Errorf("project is not compliant")
+}
+
+// LifecycleToLevelMap maps between Lifecycle and expected service maturity level
+type LifecycleToLevelMap map[int]int
+
+// Client is a opslevel client
+type Client struct {
+	*opslevel.Client // Internal opslevel client.
+
+	// LifecycleToLevelMap maps lifecycle index to level index.
+	LifecycleToLevelMap LifecycleToLevelMap
+}
+
+// newClient returns a new opslevel client configured with the token we expect to exist in
+// the environment.
+func newClient(lifecycleToLevelMap LifecycleToLevelMap) (*Client, error) {
+	opslevelToken := strings.TrimSpace(os.Getenv("OPSLEVEL_TOKEN"))
+	if opslevelToken == "" {
+		return nil, errors.New("OPSLEVEL_TOKEN environment variable is empty")
+	}
+	if len(lifecycleToLevelMap) == 0 {
+		lifecycleToLevelMap = DefaultLifecycleToLevelMap
+	}
+	return &Client{
+		opslevel.NewGQLClient(opslevel.SetAPIToken(opslevelToken)),
+		lifecycleToLevelMap,
+	}, nil
+}
+
+// isCompliant checks if the service falls within the expected maturity level.
+// This check is primarily controlled by the LifecycleToLevel map
+func (c Client) isCompliant(service *opslevel.Service, sm *opslevel.ServiceMaturity) (bool, error) {
+	if shouldBeSkipped(service.Tags.Nodes) {
+		return true, nil
+	}
+
+	currentLevelIndex := sm.MaturityReport.OverallLevel.Index
+	if len(c.LifecycleToLevelMap) < service.Lifecycle.Index {
+		return false, fmt.Errorf("unsupported lifecycle %d %s",
+			service.Lifecycle.Index, service.Lifecycle.Name)
+	}
+
+	expectedLevelIndex := c.LifecycleToLevelMap[service.Lifecycle.Index]
+	return currentLevelIndex >= expectedLevelIndex, nil
+}
+
+// shouldBeSkipped checks if a service has gating disabled. If it does, it should be skipped.
```

Comment:
I left a previous comment here but it somehow nuked itself. We should add unit tests here. Also it seems like `gating_disabled` is a specific tag for us. We should have documentation for this and probably some sort of underlying struct that we set strongly typed values for these on, that we turn `gating_disabled` into.

Reply:
This is ported from our internal `services` (an internal repo). It is tested there, but happy to duplicate the tests here too. I hate that we have to duplicate this at all, but I also wanted to make this part open.

Reply:
I disagree on making this a structure too early. I think we should wait and see what the usage looks like before settling on a structure to use.